File: jvm/periph-java/src/main/java/it/uhde/periph/chips/io_expander/Pcf8575Minimal.java
package it.uhde.periph.chips.io_expander;

import it.uhde.periph.transport.Transport;

import java.io.IOException;

public class Pcf8575Minimal {

    protected final Transport transport;
    protected int[] shadow = {0xFF, 0xFF};

    public Pcf8575Minimal(Transport transport) throws IOException {
        this.transport = transport;
        writeBoth();
    }

    private void writeBoth() throws IOException {
        transport.write(new byte[]{(byte) shadow[0], (byte) shadow[1]});
    }

    public int readPort(int port) throws IOException {
        byte[] buf = transport.read(2);
        return buf[port] & 0xFF;
    }

    public void writePort(int port, int mask) throws IOException {
        shadow[port] = mask & 0xFF;
        writeBoth();
    }

    public Pin pin(int n) {
        return new Pin(this, n);
    }

    protected void setPin(int n, boolean high) throws IOException {
        int portIdx = n / 8;
        int bit = n % 8;
        if (high) shadow[portIdx] |=   (1 << bit);
        else      shadow[portIdx] &= ~((1 << bit));
        writeBoth();
    }

    public static class Pin {
        protected final Pcf8575Minimal chip;
        protected final int n;

        protected Pin(Pcf8575Minimal chip, int n) {
            this.chip = chip;
            this.n = n;
        }

        public void setInput() throws IOException { chip.setPin(n, true); }
        public void setOutput() throws IOException { chip.setPin(n, false); }
        public void setHigh() throws IOException { chip.setPin(n, true); }
        public void setLow() throws IOException { chip.setPin(n, false); }

        public boolean read() throws IOException {
            int port = n / 8;
            int bit = n % 8;
            byte[] buf = chip.transport.read(2);
            return ((buf[port] >> bit) & 1) == 1;
        }

        public void toggle() throws IOException {
            int portIdx = n / 8;
            int bit = n % 8;
            chip.setPin(n, ((chip.shadow[portIdx] >> bit) & 1) == 0);
        }
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/io_expander/Pcf8575Minimal.kt
package it.uhde.periph.chips.io_expander

import it.uhde.periph.connection.Connection

open class Pcf8575Minimal(protected val connection: Connection) {

    var shadow = intArrayOf(0xFF, 0xFF)

    init {
        writeBoth()
    }

    internal fun writeBoth() {
        connection.write(byteArrayOf(shadow[0].toByte(), shadow[1].toByte()))
    }

    internal fun readBoth(): ByteArray = connection.read(2)

    fun readPort(port: Int): Int = readBoth()[port].toInt() and 0xFF

    fun writePort(port: Int, mask: Int) {
        shadow[port] = mask and 0xFF
        writeBoth()
    }

    open fun pin(n: Int) = Pin(this, n)

    internal fun setPin(n: Int, high: Boolean) {
        val portIdx = n / 8
        val bit = n % 8
        if (high) shadow[portIdx] = shadow[portIdx] or (1 shl bit)
        else shadow[portIdx] = shadow[portIdx] and (1 shl bit).inv()
        writeBoth()
    }

    open class Pin(val chip: Pcf8575Minimal, val n: Int) {
        fun setInput()  { chip.setPin(n, true) }
        fun setOutput() { chip.setPin(n, false) }
        fun setHigh()   { chip.setPin(n, true) }
        fun setLow()    { chip.setPin(n, false) }
        fun read(): Boolean {
            val port = n / 8
            val bit  = n % 8
            return ((chip.readBoth()[port].toInt() shr bit) and 1) == 1
        }
        fun toggle() {
            val portIdx = n / 8
            val bit = n % 8
            chip.setPin(n, ((chip.shadow[portIdx] shr bit) and 1) == 0)
        }
    }
}
